Fix unreliable logic to update site URLs registered on authentication service

benbowler opened the pull request:
## Summary

Addresses issue #2383

## Relevant technical choices

Although PHP allows comparison of two arrays using `==` I chose to compare by looping through the get_site_fields and matching each to the fetched values so that we can match for type as well as value while still ignoring the order of the items in the array.

```
foreach ( $get_site_fields as $key => $site_field ) {
	if ( ! array_key_exists( $key, $fetch_site_fields ) || $fetch_site_fields[ $key ] !== $site_field ) {
		return false;
	}
}
```

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

eugene-manuilov reviewed (changes requested):
Hi @benbowler. Is this ready for review? I don't see the following requirement from IB to be implemented. Perhaps you forgot to push your changes?

```
Update the Authentication::handle_sync_site_fields method to call $this->google_proxy->sync_site_fields( ... ) only if $this->google_proxy->are_site_fields_synced( ... ) returns FALSE;
```

benbowler commented:
You're right @eugene-manuilov, I've updated that function as per the IB.

eugene-manuilov commented on includes/Core/Authentication/Authentication.php:
```diff
@@ -1152,7 +1152,7 @@ private function handle_sync_site_fields() {
 		}
 
-		if ( $this->disconnected_reason->get() === Disconnected_Reason::REASON_CONNECTED_URL_MISMATCH ) {
+		if ( ! $this->google_proxy->are_site_fields_synced( $this->credentials ) ) {
```
@benbowler let's explicitly define that we need to sync site fields when the `are_site_fields_synced` method returns `FALSE` because it can also return an instance of the `WP_Error` class.

```suggestion
		if ( $this->google_proxy->are_site_fields_synced( $this->credentials ) === false ) {
```

eugene-manuilov commented on includes/Core/Authentication/Google_Proxy.php:
```diff
@@ -95,6 +96,75 @@ public function get_site_fields() {
+	/**
+	 * Fetch site fields
+	 *
+	 * @since x.x.x
```
```suggestion
	 * @since n.e.x.t
```

eugene-manuilov commented on includes/Core/Authentication/Google_Proxy.php:
```diff
@@ -95,6 +96,75 @@ public function get_site_fields() {
+	/**
+	 * Are site fields synced
+	 *
+	 * @since x.x.x
```
```suggestion
	 * @since n.e.x.t
```

eugene-manuilov commented on tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:
```diff
@@ -43,6 +44,134 @@ public function test_get_site_fields() {
 		);
 	}
 
+	public function test_fetch_site_fields() {
```
@benbowler let's also add a case when proxy returns an error to make sure that `fetch_site_fields` returns an appropriate WP_Error instance.

eugene-manuilov commented on tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:
```diff
@@ -43,6 +44,134 @@ public function test_get_site_fields() {
+		$success_response_data = $google_proxy->are_site_fields_synced( $credentials );
+
+		// Ensure the request was made with the proper URL and body parameters.
+		$this->assertEquals( $google_proxy->url( Google_Proxy::OAUTH2_SITE_URI ), $pre_url );
+		$this->assertEquals( 'POST', $pre_args['method'] );
+		$this->assertEqualSetsWithIndex(
+			array(
+				'site_id',
+				'site_secret',
+			),
+			array_keys( $pre_args['body'] )
+		);
```
@benbowler no need to assert functionality from the `fetch_site_fields` method. It is already checked in its own test.

eugene-manuilov commented on tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:
```diff
@@ -43,6 +44,134 @@ public function test_get_site_fields() {
+		// Ensure matching response array returns true.
+		$this->assertEquals( $success_response_data, true );
+		// Ensure WP_Error response returns WP_Error.
+		$this->assertEquals( is_wp_error( $error_response_data ), true );
```
@benbowler let's use `$this->assertWPErrorWithMessage`, defined in the `tests/phpunit/includes/TestCase.php` file, to better evaluate the error object.

eugene-manuilov reviewed (changes requested):
Thanks, @benbowler. It's a good start. Added some minor comments to address before I can approve it.

benbowler commented:
Thanks @eugene-manuilov, all updated.

eugene-manuilov commented on includes/Core/Authentication/Google_Proxy.php:
```diff
@@ -95,6 +95,72 @@ public function get_site_fields() {
+	 */
+	public function fetch_site_fields( Credentials $credentials ) {
+		if ( ! $credentials->has() ) {
+			return null;
```
@benbowler let's return an error if we don't have credentials.

eugene-manuilov commented on includes/Core/Authentication/Google_Proxy.php:
```diff
@@ -95,6 +95,72 @@ public function get_site_fields() {
+
+		$response = wp_remote_post( $this->url( self::OAUTH2_SITE_URI ), $request_args );
+
+		if ( is_wp_error( $response['body'] ) ) {
+			return $response['body'];
+		}
```
We need to check the response object itself, not its body:

```suggestion
		if ( is_wp_error( $response ) ) {
			return $response;
		}
```

benbowler replied:
This does not work because the $reponse is a standard array, the WP_Error or other results are returned in the body of the request. In the initial PR I fixed this by testing `is_wp_error($raw_body)`, we can do this or test the `$reponse['body']` let me know which you prefer.

```
var_dump($response);
```

```
array(5) {
  ["headers"]=>
  array(0) {
  }
  ["body"]=>
  object(WP_Error)#3664 (2) {
    ["errors"]=>
    array(1) {
      ["test_message"]=>
      array(1) {
        [0]=>
        string(12) "test_message"
      }
    }
    ["error_data"]=>
    array(0) {
    }
  }
  ["response"]=>
  array(2) {
    ["code"]=>
    int(200)
    ["message"]=>
    string(2) "OK"
  }
  ["cookies"]=>
  array(0) {
  }
  ["http_response"]=>
  NULL
}
```

eugene-manuilov replied:
@benbowler see the documentation for the [`wp_remote_post`](https://developer.wordpress.org/reference/functions/wp_remote_post/) function:

```
Return #
(array|WP_Error) The response or WP_Error on failure.
```

eugene-manuilov reviewed (changes requested):
Thanks, @benbowler! Just two more minor things to address and I'll approve it.

benbowler commented:
Hey @eugene-manuilov, I've updated the WP_Error for oauth_credentials_not_exist. For the other request, it's not possible to check the raw `$response` array `is_wp_error`, I've left a comment above and reverted to my original implementation of checking `$raw_body` `is_wp_error`.

benbowler commented:
Updated.

benbowler commented:
@eugene-manuilov

eugene-manuilov commented on includes/Core/Authentication/Google_Proxy.php:
```diff
@@ -95,6 +95,72 @@ public function get_site_fields() {
+	 */
+	public function fetch_site_fields( Credentials $credentials ) {
+		if ( ! $credentials->has() ) {
+			return new WP_Error( 'oauth_credentials_not_exist' );
```
@benbowler, please, add `use WP_Error;` to the top of the class to make it work here.

eugene-manuilov commented on includes/Core/Authentication/Google_Proxy.php:
```diff
@@ -95,6 +95,72 @@ public function get_site_fields() {
+			return $response;
+		}
+
+		$raw_body = wp_remote_retrieve_body( $response );
```
@benbowler let's also make sure that the response status code is 200:

```php
if ( wp_remote_retrieve_response_code( $response ) != 200 ) {
    return new WP_Error( ... );
}

$raw_body = wp_remote_retrieve_body( $response );
```

eugene-manuilov commented on tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:
```diff
@@ -43,6 +44,121 @@ public function test_get_site_fields() {
+		);
+
+		// Force WP_Error response from as http requests are blocked.
+		$this->mock_http_request(
+			$google_proxy->url( 'http://example.com' ),
+			null
+		);
```
@benbowler let's better use the real URL for it. You can create another helper function to test errors. Something like this:

```php
private function mock_http_failure( $request_url, $response_error ) {
    add_filter(
        'pre_http_request',
        function( $response, $parsed_args, $url ) use ( $request_url, $response_error ) {
            if ( $url === $request_url ) {
                return new WP_Error( $response_error );
            } else {
                return $response;
            }
        },
        10,
        3
    );
}
```

eugene-manuilov reviewed (changes requested):
Thanks, @benbowler. Found a few more things to improve. Could you please check it?

eugene-manuilov reviewed (dismissed):
LGTM! Thanks, @benbowler!

felixarntz commented on includes/Core/Authentication/Google_Proxy.php:
```diff
@@ -95,6 +96,76 @@ public function get_site_fields() {
+			return $response;
+		}
+
+		if ( wp_remote_retrieve_response_code( $response ) !== 200 ) {
+			return new WP_Error( 'non_200_response_code' );
+		}
```
This is somewhat unprecise, as the `WP_Error` returned would not provide information on the real error here - which may be okay as a last resort, but we should try to use more specific errors. For example, below you run the check whether the response contains an `error` key, which is possible, and in that case the response status code would not be 200. So we'd want to in such a case return the code in `error` rather than a less precise error that the response didn't come back with a 200.

I think we should just remove this clause completely. If the entire request fails, that would be caught by the `WP_Error` from `wp_remote_post`, and if the request comes back with an error response, we should be okay relying on the `error` key in the response.

benbowler replied:
Great, updated and pushed in the most recent push.

felixarntz commented on tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:
```diff
@@ -218,20 +334,24 @@ public function test_exchange_site_code() {
 	 * given URL will yield a specific response.
 	 *
 	 * @param string $request_url   Request URL to modify response for.
-	 * @param array  $response_data Response data to return for the request. Will be JSON-encoded.
+	 * @param (array)  $response_data Response data to return for the request. Will be JSON-encoded if it is an array.
 	 * @param int    $response_code Optional. Response status code to return. Default 200.
 	 */
-	private function mock_http_request( $request_url, array $response_data, $response_code = 200 ) {
+	private function mock_http_request( $request_url, $response_data, $response_code = 200 ) {
 		add_filter(
 			'pre_http_request',
 			function ( $preempt, $args, $url ) use ( $request_url, $response_data, $response_code ) {
 				if ( $request_url !== $url ) {
 					return $preempt;
 				}
 
+				if ( is_array( $response_data ) ) {
+					$response_data = json_encode( $response_data );
+				}
+
 				return array(
 					'headers'       => array(),
-					'body'          => json_encode( $response_data ),
+					'body'          => $response_data,
```
Why was this changed? I don't see any usage where `$response_data` is not an array, and I think it's better to keep the signature consistent, especially if there's no reason to change it.

benbowler replied:
In a previous commit, I updated this function to allow tests to pass a WP_Error via $response_data before splitting it out later into `mock_http_failure`.

I've added this type back in. 👌

felixarntz commented on tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:
```diff
@@ -244,4 +364,26 @@ function ( $preempt, $args, $url ) use ( $request_url, $response_data, $response
+	 * Adds a 'pre_http_request' filter that will ensure the request for the
+	 * given URL will return WP_Error.
+	 *
+	 * @param string $request_url   Request URL to modify response for.
+	 * @param (WP_Error)  $response_error Response WP_Error to return.
```
This shouldn't be in parentheses, and the parameter docs should be aligned.
```suggestion
	 * @param string   $request_url    Request URL to modify response for.
	 * @param WP_Error $response_error Response WP_Error to return.
```

felixarntz reviewed (changes requested):
Great! Just some minor feedback on things that could be improved.

Normally I wouldn't block merging on such changes, but since this is still awaiting for the authentication service side to be launched, I think it's worth making these minor enhancements.

felixarntz commented on tests/phpunit/integration/Core/Authentication/Google_ProxyTest.php:
```diff
@@ -229,9 +345,13 @@ function ( $preempt, $args, $url ) use ( $request_url, $response_data, $response
 					return $preempt;
 				}
 
+				if ( is_array( $response_data ) ) {
+					$response_data = json_encode( $response_data );
+				}
```
The check here is also no longer necessary, since it'll always be an array. We can unconditionally `json_encode` it like before.

felixarntz reviewed (approved):
Awesome! Logic on our end is now pushed to the staging version of the authentication service, and I've updated the QA Brief with the related information for testing.
